### DISCUSSION TITLE
Using Windows App SDK in Qt app (without a Visual Studio project)

### DISCUSSION THREAD

**Discussion by triplef** [microsoft/WindowsAppSDK#1891] [Q&A] [answered]
We’re building our packaged Win32 app using Qt and would like to use the Windows App SDK. Is this a supported scenario or has anyone managed to use the Windows App SDK / WinUI outside of a Visual Studio project? Specifically, we’d like to take advantage of the AppWindow API to make our app window look more native/modern.

I downloaded the [Microsoft.WindowsAppSDK Nuget package](https://www.nuget.org/packages/Microsoft.WindowsAppSDK/), but it seems to be missing most of the headers for WinUI or the AppWindow API. I read that these are provided as sub-projects to the Nuget package – is there any way to download these manually?

**Comment by BenJKuhn** [comment]
We've focused primarily on .NET and msbuild up to this point, but have been collecting feedback related to supporting other build environments in the future. Are you developing in C++ then ? Which build system are you using?

Thanks,

Ben

**Reply by triplef** [reply]
Thanks for your response! Yes we are developing in C++ and are using Qt’s QMake as a build system, although we may switch to CMake at some point as that is what Qt is migrating to as a whole.

We’re comfortable doing the integration into our build system ourselves as long as we have a way to acquire all the required libraries and headers and there’s no additional extensive tooling requirements. We’d appreciate if this could be made possible to broaden the possible use of WinUI.

**Reply by DrusTheAxe** [reply]
> as long as we have a way to acquire all the required libraries and headers

The NuGet contains all the bits needed at build-time. I'm not very familiar with CMake but I'd expect e.g. adding headers to the include path to be pretty straightforward. Do the .props and .targets MSBuild logic work under CMake or do they need be ported/converted?

P.S. The nuget also contains the MSIX packages, needed to install for use at runtime. Unless you prefer to download and run the installer.

>We’d appreciate if this could be made possible to broaden the possible use of WinUI.

Our [productboard portal](https://portal.productboard.com/winappsdk/1-windows-app-sdk) doesn't have an item for CMake support. Please hop over there and submit one to [ensure it's on our radar](https://github.com/microsoft/WindowsAppSDK/discussions/1778). Thank you

**Reply by wjk** [reply]
> Do the .props and .targets MSBuild logic work under CMake or do they need be ported/converted?

In a word, no. The logic would have to be rewritten to work under CMake. The fact that there is still no standalone tool to generate xbf files makes this quite difficult.

**Reply by triplef** [reply]
> The NuGet contains all the bits needed at build-time. 

The Microsoft.WindowsAppSDK NuGet package does not seem to contain headers for WinUI (e.g. `winrt/Windows.UI.WindowManagement.h`). Are there additional packages containing these?

> The fact that there is still no standalone tool to generate xbf files makes this quite difficult.

Would we need to generate such files? We just want to access e.g. the [AppWindow API](https://docs.microsoft.com/en-us/uwp/api/windows.ui.windowmanagement.appwindow?view=winrt-22000) from C++/WinRT code.

**Reply by wjk** [reply]
XBF files are only needed if you are writing your application using WinUI 3. It is actually quite easy to reference the rest of AppSDK from any build system. (I also think you meant to link to [this AppWindow](https://docs.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.windowing.appwindow?view=windows-app-sdk-1.0) instead.)

**Reply by triplef** [reply]
Ah yes, thanks for the correction on the API link.

Can you elaborate on how to reference the SDK from any build system? Where do we get the required libraries and header files if they are not in the NuGet package?

**Reply by wjk** [reply]
> Where do we get the required libraries and header files if they are not in the NuGet package?

They are in the NuGet package. While you cannot consume it as intended unless you're using MSBuild, remember that a NuGet package is just a zip file underneath. You can extract its contents just like any other zip file, and then you would have loose files on disk like you are requesting.

**Reply by triplef** [reply]
> They are in the NuGet package. While you cannot consume it as intended unless you're using MSBuild, remember that a NuGet package is just a zip file underneath. You can extract its contents just like any other zip file, and then you would have loose files on disk like you are requesting.

Yes I realize that, but if you unzip the [Microsoft.WindowsAppSDK NuGet package](https://www.nuget.org/packages/Microsoft.WindowsAppSDK/) it seems to be missing most of the headers
(e.g. `winrt/Microsoft.UI.Windowing.h` required for AppWindow). In fact, the `include/winrt` directory only contains two header files: `Microsoft.UI.Interop.h` and `Microsoft.UI.Composition.Interop.h`. What am I missing?

**Reply by DrusTheAxe** [reply]
How are you consuming WinRT APIs? C++/WinRT? WRL? Other?

WinAppSDK's NuGet only contains *.winmd. That suffices for some consumers (e.g. C++/WinRT), for others there are various tools and techniques to generate derivative artifacts (e.g. *.h) from *.winmd. I defer to @BenJKuhn for which are best practice these days

**Reply by triplef** [reply]
We're using C++/WinRT with the Clang/LLVM toolchain (which is fairly compatible with MSVC).

I'm guessing for this setup we'll have to generate header files from the WinMD files since we're not using Visual Studio?

**Reply by wjk** [reply]
> I'm guessing for this setup we'll have to generate header files from the WinMD files since we're not using Visual Studio?

That’s correct. VS invokes cppwinrt.exe automatically to generate the header files; it can also be run manually.

**Reply by triplef** [reply]
Thanks @wjk and everyone else for your help, I'll give cppwinrt.exe a try.

For anyone else following along this page has a bit more info about the process and compiler options that might be required:
https://docs.microsoft.com/en-us/windows/uwp/cpp-and-winrt-apis/intro-to-using-cpp-with-winrt#visual-studio-support-for-cwinrt-xaml-the-vsix-extension-and-the-nuget-package

**Reply by triplef** [reply]
So I was able to generate the header files by running this inside the unpacked Microsoft.WindowsAppSDK.1.0.0 package folder:

```
cppwinrt.exe -in lib\uap10.0.18362 -out include -ref 10.0.22000.0
```

I then linked against `Microsoft.WindowsAppRuntime.lib`, `Microsoft.WindowsAppRuntime.Bootstrap.lib`, and `WindowsApp.lib`, added a call to [`MddBootstrapInitialize()`](https://docs.microsoft.com/en-us/windows/apps/windows-app-sdk/tutorial-unpackaged-deployment?tabs=cpp) in my main, and finally added this code to my main window implementation:

```c
auto windowId = winrt::Microsoft::UI::GetWindowIdFromWindow(hWnd);
auto appWindow = winrt::Microsoft::UI::Windowing::AppWindow::GetFromWindowId(windowId);
appWindow.Title(L"WinUI Title");
```

The app now builds and links fine, but I am getting the following exception in the call to `GetFromWindowId()`:

> 0x00000505 (ERROR_DELAY_LOAD_FAILED): An attempt to delay-load a .dll or get a function address in a delay-loaded .dll failed (parameters: 0x00007FFDAFB97640, 0x00007FFDAFBAEEAE).

Any ideas what might be the issue here?

**Reply by riverar** [reply]
You'll need to run this in a debugger, ideally with loader snaps enabled, to root that out. But it's likely either Microsoft.WindowsAppRuntime.Bootstrap.dll doesn't exist on disk next to your executable or you don't have the appropriate [VC runtimes ](https://docs.microsoft.com/en-us/cpp/windows/latest-supported-vc-redist?view=msvc-170) installed. I'd be happy to help you debug it, if needed. (Drop a zip somewhere or [email it](mailto:rafael[remove]@withinwindows.com).)

**Reply by triplef** [reply]
Thanks for the quick response @riverar! I do have `Microsoft.WindowsAppRuntime.Bootstrap.dll` next to my exe, and I think I have the VC runtimes.

Here is what I get with Loader Snaps (after running `gflags /i WinUIQtTest.exe +sls`) – looks like `ext-ms-win-windowing-internal-l1-1-0.dll` is missing. Any thoughts where I would get that one?

```
...
0c50:34dc @ -2009611312 - LdrpGetProcedureAddress - INFO: Locating procedure "DllGetActivationFactory" by name
0c50:34dc @ -2009611312 - LdrpLoadDllInternal - ENTER: DLL name: ext-ms-win-windowing-internal-l1-1-0.dll
0c50:34dc @ -2009611312 - LdrpFindKnownDll - ENTER: DLL name: ext-ms-win-windowing-internal-l1-1-0.dll
0c50:34dc @ -2009611312 - LdrpFindKnownDll - RETURN: Status: 0xc0000135
0c50:34dc @ -2009611312 - LdrpSearchPath - ENTER: DLL name: ext-ms-win-windowing-internal-l1-1-0.dll
0c50:34dc @ -2009611312 - LdrpComputeLazyDllPath - INFO: DLL search path computed: C:\Program Files\WindowsApps\Microsoft.WindowsAppRuntime.1.0_0.319.455.0_x64__8wekyb3d8bbwe;C:\Windows\SYSTEM32;C:\Windows\system;C:\Windows;.;C:\Program Files\WindowsApps\Microsoft.WindowsAppRuntime.1.0_0.319.455.0_x64__8wekyb3d8bbwe;C:\Python310\Scripts\;C:\Python310\;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Windows\System32\OpenSSH\;C:\Program Files\Git LFS;C:
0c50:34dc @ -2009611312 - LdrpResolveDllName - ENTER: DLL name: C:\Program Files\WindowsApps\Microsoft.WindowsAppRuntime.1.0_0.319.455.0_x64__8wekyb3d8bbwe\ext-ms-win-windowing-internal-l1-1-0.dll
0c50:34dc @ -2009611312 - LdrpResolveDllName - RETURN: Status: 0xc0000135
0c50:34dc @ -2009611312 - LdrpResolveDllName - ENTER: DLL name: C:\Windows\SYSTEM32\ext-ms-win-windowing-internal-l1-1-0.dll
0c50:34dc @ -2009611312 - LdrpResolveDllName - RETURN: Status: 0xc0000135
0c50:34dc @ -2009611312 - LdrpResolveDllName - ENTER: DLL name: C:\Windows\system\ext-ms-win-windowing-internal-l1-1-0.dll
0c50:34dc @ -2009611312 - LdrpResolveDllName - RETURN: Status: 0xc0000135
0c50:34dc @ -2009611312 - LdrpResolveDllName - ENTER: DLL name: C:\Windows\ext-ms-win-windowing-internal-l1-1-0.dll
0c50:34dc @ -2009611312 - LdrpResolveDllName - RETURN: Status: 0xc0000135
0c50:34dc @ -2009611312 - LdrpResolveDllName - ENTER: DLL name: .\ext-ms-win-windowing-internal-l1-1-0.dll
0c50:34dc @ -2009611312 - LdrpResolveDllName - RETURN: Status: 0xc0000135
0c50:34dc @ -2009611312 - LdrpResolveDllName - ENTER: DLL name: C:\Program Files\WindowsApps\Microsoft.WindowsAppRuntime.1.0_0.319.455.0_x64__8wekyb3d8bbwe\ext-ms-win-windowing-internal-l1-1-0.dll
0c50:34dc @ -2009611312 - LdrpResolveDllName - RETURN: Status: 0xc0000135
0c50:34dc @ -2009611312 - LdrpResolveDllName - ENTER: DLL name: C:\Python310\Scripts\ext-ms-win-windowing-internal-l1-1-0.dll
0c50:34dc @ -2009611312 - LdrpResolveDllName - RETURN: Status: 0xc0000135
0c50:34dc @ -2009611312 - LdrpResolveDllName - ENTER: DLL name: C:\Python310\ext-ms-win-windowing-internal-l1-1-0.dll
0c50:34dc @ -2009611312 - LdrpResolveDllName - RETURN: Status: 0xc0000135
0c50:34dc @ -2009611312 - LdrpResolveDllName - ENTER: DLL name: C:\Windows\system32\ext-ms-win-windowing-internal-l1-1-0.dll
0c50:34dc @ -2009611312 - LdrpResolveDllName - RETURN: Status: 0xc0000135
0c50:34dc @ -2009611312 - LdrpResolveDllName - ENTER: DLL name: C:\Windows\ext-ms-win-windowing-internal-l1-1-0.dll
0c50:34dc @ -2009611312 - LdrpResolveDllName - RETURN: Status: 0xc0000135
0c50:34dc @ -2009611312 - LdrpResolveDllName - ENTER: DLL name: C:\Windows\System32\Wbem\ext-ms-win-windowing-internal-l1-1-0.dll
0c50:34dc @ -2009611312 - LdrpResolveDllName - RETURN: Status: 0xc0000135
0c50:34dc @ -2009611312 - LdrpResolveDllName - ENTER: DLL name: C:\Windows\System32\WindowsPowerShell\v1.0\ext-ms-win-windowing-internal-l1-1-0.dll
0c50:34dc @ -2009611312 - LdrpResolveDllName - RETURN: Status: 0xc0000135
0c50:34dc @ -2009611312 - LdrpResolveDllName - ENTER: DLL name: C:\Windows\System32\OpenSSH\ext-ms-win-windowing-internal-l1-1-0.dll
0c50:34dc @ -2009611312 - LdrpResolveDllName - RETURN: Status: 0xc0000135
0c50:34dc @ -2009611312 - LdrpResolveDllName - ENTER: DLL name: C:\Program Files\Git LFS\ext-ms-win-windowing-internal-l1-1-0.dll
0c50:34dc @ -2009611312 - LdrpResolveDllName - RETURN: Status: 0xc0000135
0c50:34dc @ -2009611312 - LdrpResolveDllName - ENTER: DLL name: C:\ProgramData\chocolatey\bin\ext-ms-win-windowing-internal-l1-1-0.dll
0c50:34dc @ -2009611312 - LdrpResolveDllName - RETURN: Status: 0xc0000135
0c50:34dc @ -2009611312 - LdrpResolveDllName - ENTER: DLL name: C:\Program Files\Git\cmd\ext-ms-win-windowing-internal-l1-1-0.dll
0c50:34dc @ -2009611312 - LdrpResolveDllName - RETURN: Status: 0xc0000135
0c50:34dc @ -2009611312 - LdrpResolveDllName - ENTER: DLL name: C:\Program Files\dotnet\ext-ms-win-windowing-internal-l1-1-0.dll
0c50:34dc @ -2009611312 - LdrpResolveDllName - RETURN: Status: 0xc0000135
0c50:34dc @ -2009611312 - LdrpResolveDllName - ENTER: DLL name: C:\Users\frede\AppData\Local\Microsoft\WindowsApps\ext-ms-win-windowing-internal-l1-1-0.dll
0c50:34dc @ -2009611312 - LdrpResolveDllName - RETURN: Status: 0xc0000135
0c50:34dc @ -2009611312 - LdrpResolveDllName - ENTER: DLL name: C:\Users\frede\Qt\5.15.7\msvc2019_64\bin\ext-ms-win-windowing-internal-l1-1-0.dll
0c50:34dc @ -2009611312 - LdrpResolveDllName - RETURN: Status: 0xc0000135
0c50:34dc @ -2009611312 - LdrpResolveDllName - ENTER: DLL name: C:\Users\frede\AppData\Local\atom\bin\ext-ms-win-windowing-internal-l1-1-0.dll
0c50:34dc @ -2009611312 - LdrpResolveDllName - RETURN: Status: 0xc0000135
0c50:34dc @ -2009611312 - LdrpResolveDllName - ENTER: DLL name: C:\tools\msys64\ext-ms-win-windowing-internal-l1-1-0.dll
0c50:34dc @ -2009611312 - LdrpResolveDllName - RETURN: Status: 0xc0000135
0c50:34dc @ -2009611312 - LdrpResolveDllName - ENTER: DLL name: C:\Users\frede\.dotnet\tools\ext-ms-win-windowing-internal-l1-1-0.dll
0c50:34dc @ -2009611312 - LdrpResolveDllName - RETURN: Status: 0xc0000135
0c50:34dc @ -2009611312 - LdrpSearchPath - RETURN: Status: 0xc0000135
0c50:34dc @ -2009611312 - LdrpProcessWork - ERROR: Unable to load DLL: "ext-ms-win-windowing-internal-l1-1-0.dll", Parent Module: "C:\Program Files\WindowsApps\Microsoft.WindowsAppRuntime.1.0_0.319.455.0_x64__8wekyb3d8bbwe\Microsoft.UI.Windowing.Core.dll", Status: 0xc0000135
0c50:34dc @ -2009611312 - LdrpLoadDllInternal - RETURN: Status: 0xc0000135
0c50:34dc @ -2009611312 - LdrpRedirectDelayloadFailure - ERROR: Failed to find export ext-ms-win-windowing-internal-l1-1-0.dll!Windowing_DllGetActivationFactory (Ordinal:0) in "Microsoft.UI.Windowing.Core.dll"  0xc0000135
Exception thrown at 0x00007FFE3FAC4F69 (KernelBase.dll) in WinUIQtTest.exe: 0x00000505: An attempt to delay-load a .dll or get a function address in a delay-loaded .dll failed (parameters: 0x00007FFDC7A47640, 0x00007FFDC7A5EEAE).
0c50:34dc @ -2009611312 - LdrpGetProcedureAddress - INFO: Locating procedure "FlsGetValue" by name
Unhandled exception at 0x00007FFE3FAC4F69 (KernelBase.dll) in WinUIQtTest.exe: 0x00000505: An attempt to delay-load a .dll or get a function address in a delay-loaded .dll failed (parameters: 0x00007FFDC7A47640, 0x00007FFDC7A5EEAE).
```

**Reply by riverar** [reply]
`ext-ms-win-windowing-internal-l1-1-0.dll` is not a real library but rather an _API set_ that should eventually point to `Microsoft.UI.Windowing.Core.dll`. You mentioned you already called `MddBootstrapInitialize` -- did it succeed?

**Reply by triplef** [reply]
Yes, `MddBootstrapInitialize()` returns 0. This is the code I run in my main:

```c
const UINT32 majorMinorVersion{ 0x00010000 };
PCWSTR versionTag{ L"" };
const PACKAGE_VERSION minVersion{};
const HRESULT hr{ MddBootstrapInitialize(majorMinorVersion, versionTag, minVersion) };
```

**Reply by DrusTheAxe** [reply]
>ERROR: Failed to find export ext-ms-win-windowing-internal-l1-1-0.dll!Windowing_DllGetActivationFactory (Ordinal:0) in "Microsoft.UI.Windowing.Core.dll"

What OS are you running on?

`Microsoft.UI.Windowing.Core.dll` imports `Windowing_DllGetActivationFactory()` from ext-ms-win-windowing-internal-l1-1-0.dll (it's the only symbol imported from it). That APIset(1) doesn't exist downlevel. I just verified it doesn't exist on 21H1 (my home PC).

WinAppSDK supports down to RS5 so this is a bug. Even if Windowing isn't supported downlevel(2) it should gracefully degrade on older systems, not crash.

@rkarman I filed https://task.ms/37687436 and marked it for consideration for servicing.

As a temporary partial workaround you can check if it exists before calling GetWindowIdFromWindow() e.g.

```
inline bool IsGetWindowIdFromWindowSupported()
{
    HANDLE h{ LoadLibraryW(L"ext-ms-win-windowing-internal-l1-1-0.dll") };
    const auto exists{ !h };
    if (h)
    {
        FreeLibrary(h);
    }
    return exists;
}
```

or if you use WIL

```
inline bool IsGetWindowIdFromWindowSupported()
{
    wil::unique_hmodule dll{ LoadLibraryW(L"ext-ms-win-windowing-internal-l1-1-0.dll") };
    return !!dll;
}
```

That won't make the API magically work on older systems but at least you can check that before use and not crash.

(1) APIsets are code that look like DLLs (LoadLibrary works, export symbols, etc) but they're virtual not physical files, their code hosted by some actual file. APIsets were invented in Vista as a way to refactor Windows' APIs, move code around to different hosting files to better compose Windows (for layering reasons, or performance/efficiency, or....) without breaking the zillions of existing callers that expect Foo.dll to export Bar(). As usual, all problems are solved by another layer of indirection :-)

That ext-* is an extension APIset, which support with a form of delayload support (similar to VS's /DELAYLOAD, but implemented in the OS). The code in Microsoft.UI.Windowing.Core.dll calling `Windowing_DllGetActivationFactory()` isn't doing checking if the extension APIset's present before use and thus boom :-(

(2) WinAppSDK supports Windows down to RS5. Some functionality isn't supported all the way down to that (pen support is one example) but that's more the exception than the rule.

**Reply by triplef** [reply]
Thanks for the explanations @DrusTheAxe!

I’m running the latest official Windows 10 release 21H2, so it sounds like the API set should be there. Additionally I can run the Windowing sample project, which also [calls `GetWindowIdFromWindow()`](https://github.com/microsoft/WindowsAppSDK-Samples/blob/main/Samples/Windowing/cpp-winui/SampleApp/MainWindow.xaml.cpp#L92), without issues on my system.

Is there anything else I could be missing that Visual Studio does behind the scenes, or do I need to maybe call cppwinrt with different parameters than the following? I didn’t find much too documentation on it and ran with the first thing that let me build my code...

`cppwinrt.exe -in lib\uap10.0.18362 -out include -ref 10.0.22000.0`

**Reply by DrusTheAxe** [reply]
>I’m running the latest official Windows 10 release 21H2, so it sounds like the API set should be there.

I didn't say that, just that it's not present on <=21H1. I'm not sure when it first appeared. @ionvasilian would know more

>do I need to maybe call cppwinrt with different parameters than the following

That looks OK by me though IANACWE (I Am Not A C++/WinRT Expert). But I don't see how that would matter, this error should occur if GetWindowIdFromWindow is called (for any reason) given the right platform / environmental soup. It's been escalated to the windowing team for further investigation.

If you use VS the C++/WinRT pluigin (vsix?) integrates into VS so you get the usual property UI under project Settings to fiddle with values, build pipeline integration etc. If you need to know more C++/WinRT experts are pretty active on the github project pages

**Reply by triplef** [reply]
I’ve created a small sample project with Visual Studio 2022 (17.0.4) that reproduces the issue for me:
https://www.dropbox.com/s/16qq0uq28vq35m9/WinUITest.zip?dl=0

Here is how I set up the project:

1. Create a new VS project using the "Windows Desktop Application" template
2. Add Microsoft.WindowsAppSDK NuGet package
3. Generate headers from the WinMD files by running the following inside the Windows App SDK package folder:
   `cppwinrt.exe -in lib\uap10.0.18362 -out ..\..\WindowsAppSDK-include-gen -ref 10.0.22000.0`
4. Add the folder with the generated headers to the include path and configure project to use C++17.
5. Call `MddBootstrapInitialize()` in `wWinMain()`, and add some code to call `GetWindowIdFromWindow()` in `InitInstance()`

Please let me know if there’s anything else I can do to help track down this issue.

**Reply by DrusTheAxe** [reply]
We believe we've identified the root cause<sup>1</sup> and working on a fix. Once verified it'll be considered for 1.0-servicing and very likely to meet the bar.

<sup>1</sup> The logic detecting if a new(ish) function is available potentially got wrong answers due to specific code (some N layers deep) not working as expected ('cause, complex compatibility reasons). An alternative implementation that doesn't hit this problem is being pursued.

Thank you for reporting this, and for your patience and assistance helping us identify the problem.

+@jeffstall  @tiberiu2  @ionvasilian for future updates on the fix

**Reply by gabbybilka** [reply]
Fix is in 1.0.1 servicing: #2291

**Comment by triplef** [comment, marked as answer]
### Solution for unpackaged app

Notwithstanding the issue with `GetWindowIdFromWindow()` discussed in the other thread, we managed to integrate the Windows App SDK in an **unpackaged app** without using Visual Studio as follows:

- Install Windows App SDK:  
  ```batch
  :: download Windows App SDK NuGet package
  nuget.exe install Microsoft.WindowsAppSDK -OutputDirectory %cd%
  
  :: generate headers from Windows Metadata files
  cppwinrt.exe -out include-gen -ref sdk -in Microsoft.WindowsAppSDK.1.0.0\lib\uap10.0 -in Microsoft.WindowsAppSDK.1.0.0\lib\uap10.0.17763 -in Microsoft.WindowsAppSDK.1.0.0\lib\uap10.0.18362
  ```
- Add `Microsoft.WindowsAppSDK.1.0.0/include` and `include-gen` to header search paths.
- Link against `Microsoft.WindowsAppRuntime.lib` and `Microsoft.WindowsAppRuntime.Bootstrap.lib`.
- Bundle `Microsoft.WindowsAppRuntime.Bootstrap.dll` with app.
- Call `MddBootstrapInitialize()` at runtime as [explained here](https://learn.microsoft.com/en-us/windows/apps/windows-app-sdk/tutorial-unpackaged-deployment?tabs=cpp).

**Comment by triplef** [comment]
### Deployment issues in packaged app

We actually want to use this in a **packaged app**, but we’re running into some issues with the deployment following [this documentation](https://docs.microsoft.com/en-us/windows/apps/windows-app-sdk/deploy-packaged-apps). Here is what we did so far:

- We added a package dependency to `Microsoft.WindowsAppRuntime.1.0` in our `AppxManifest.xml`:
  ```xml
  <Dependencies>
    <TargetDeviceFamily Name="Windows.Desktop" MinVersion="10.0.18362.0" MaxVersionTested="10.0.19044.0" />
    <PackageDependency Name="Microsoft.VCLibs.140.00.UWPDesktop" MinVersion="14.0.24217.0" Publisher="CN=Microsoft Corporation, O=Microsoft Corporation, L=Redmond, S=Washington, C=US" />
    <PackageDependency Name="Microsoft.WindowsAppRuntime.1.0" MinVersion="0.318.928.0" Publisher="CN=Microsoft Corporation, O=Microsoft Corporation, L=Redmond, S=Washington, C=US" />
  </Dependencies>
  ```
- At runtime, we’re calling `DeploymentManager::GetStatus()`, which returns `DeploymentStatus::PackageInstallRequired`.
- So we call `DeploymentManager::Initialize()`, which returns `DeploymentStatus::PackageInstallFailed` with extended error 0x800700E8 (ERROR_NO_DATA).

I am guessing that we are again missing some other step performed by Visual Studio.

@DrusTheAxe any idea what we could be missing here?

**Reply by triplef** [reply]
I’ve posted this as its own follow-up discussion: https://github.com/microsoft/WindowsAppSDK/discussions/2139.

**Reply by DrusTheAxe** [reply]
@sachintaMSFT for follow up re the `ERROR_NO_DATA`
